pymanopt/tools/diagnostics.py

```diff
@@ -0,0 +1,157 @@
+    This function attempts to identify a contiguous segment of the curve
+    defined by the vectors x and y that appears to be linear. A line is fit
+    through the data over all windows of length window_length and the best
+    fit is retained. The output specifies the range of indices such that
+    x(segment) is the portion over which (x, y) is the most linear and the
+    output poly specifies a first order polynomial that best fits (x, y) over
+    that segment (highest degree coefficients first).
+    See also: check_directional_derivative check_gradient
+    """
+    residues = np.zeros(len(x)-window_length)
+    polys = np.zeros(shape=(2, len(residues)))
+    for i in range(len(residues)):
+        segment = range(i, (i+window_length)+1)
+        poly, residuals, _, _, _ = np.polyfit(x[segment], y[segment],
+                                              1, full=True)
+        residues[i] = np.linalg.norm(residuals)
+        polys[:, i] = poly
+    best = np.argmin(residues)
+    segment = range(best, best+window_length+1)
+    poly = polys[:, best]
+    return segment, poly
+
+
+def check_directional_derivative(problem, x=None, d=None):
+    """Checks the consistency of the cost function and directional derivatives.
+    check_directional_derivative performs a numerical test to check that the
+    directional derivatives defined in the problem structure agree up to first
+    order with the cost function at some point x, along some direction d. The
+    test is based on a truncated Taylor series (see online pymanopt
+    documentation).
+    Both x and d are optional and will be sampled at random if omitted.
+    See also: check_gradient
+    """
+    #  If x and / or d are not specified, pick them at random.
+    if d is not None and x is None:
+        raise ValueError("If d is provided, x must be too, "
+                         "since d is tangent at x.")
+    if x is None:
+        x = problem.manifold.rand()
+    if d is None:
+        d = problem.manifold.randvec(x)
+
+    # Compute the value f0 at f and directional derivative at x along d.
+    f0 = problem.cost(x)
+    grad = problem.grad(x)
+    df0 = problem.manifold.inner(x, grad, d)
+
+    # Compute the value of f at points on the geodesic (or approximation
+    # of it) originating from x, along direction d, for stepsizes in a
+    # large range given by h.
+    h = np.logspace(-8, 0, 51)
+    value = np.zeros_like(h)
+    for i, h_k in enumerate(h):
+        try:
+            y = problem.manifold.exp(x, h_k * d)
+        except NotImplementedError:
+            y = problem.manifold.retr(x, h_k * d)
+        value[i] = problem.cost(y)
+
+    # Compute the linear approximation of the cost function using f0 and
+    # df0 at the same points.
+    model = np.polyval([df0, f0], h)
+
+    # Compute the approximation error
+    err = np.abs(model - value)
+    model_is_exact = not np.all(err < 1e-12)
+    if model_is_exact:
```

Comment:
The conditions are inverted now. The line needs to read `if not model_is_exact`, or the conditional bodies should be inverted, which I personally would prefer.

Reply:
Ok, I swapped the conditional bodies. I completely agree, it is more readable.